File: abt.ml
(* -*- mode: ocaml; -*- *)
(* A copy of https://gist.github.com/neel-krishnaswami/834b892327271e348f79 *)

module type FUNCTOR = sig
  type 'a t
  val map : ('a -> 'b) -> 'a t -> 'b t
end

type 'a monoid = {unit : 'a ; join : 'a -> 'a -> 'a}

type var = string
module V = Set.Make(struct type t = var let compare = compare end)

module type ABT = sig
  type 'a signature
  type 'a f = Var of var | Abs of var * 'a | Tm of 'a signature
  val map : ('a -> 'b) -> 'a f -> 'b f

  type t
  val into : t f -> t
  val out  : t -> t f

  val freevars : t -> V.t
  val var : V.elt -> t
  val abs : V.elt * t -> t
  val tm : t signature -> t

  val subst : t -> var -> t -> t
end

module type SIGNATURE = sig
  include FUNCTOR

  val join : 'a monoid -> 'a t -> 'a
end

module Abt(F : SIGNATURE) : ABT with type 'a signature := 'a F.t =
struct
  type 'a f =
    | Var of var
    | Abs of var * 'a
    | Tm of 'a F.t

  let map f = function
    | Var x -> Var x
    | Abs(x, e) -> Abs(x, f e)
    | Tm t -> Tm (F.map f t)

  type t = In of V.t * t f

  let freevars (In(vs, _)) = vs
  let out (In(_, t)) = t

  let m = {unit = V.empty; join = V.union}

  let var x = In(V.singleton x, Var x)
  let abs(z, e) = In(V.remove z (freevars e), Abs(z, e))
  let tm t = In(F.join m (F.map freevars t), Tm t)

  let into = function
    | Var x -> var x
    | Abs(x, e) -> abs(x, e)
    | Tm t -> tm t

  let rec fresh vs v =
    if V.mem v vs then fresh vs (v ^ "'") else v

  let rec rename x y (In(fvs, t)) =
    match t with
    | Var z -> if x = z then var y else var z
    | Abs(z, e) -> if x = z then abs(z, e) else abs(z, rename x y e)
    | Tm v -> tm (F.map (rename x y) v)

  let rec subst t x body =
    match out body with
    | Var z when x = z -> t
    | Var z            -> var z
    | Abs(x, e) ->
      let x' = fresh (V.union (freevars t) (freevars body)) x in
      let e' = subst t x (rename x x' e) in
      abs(x', e')
    | Tm body -> tm (F.map (subst t x) body)
end

module Lambda =
struct
  type tp = Base | Arrow of tp * tp
  type 'a t = Lam of 'a | App of 'a * 'a | Let of 'a * 'a | Annot of tp * 'a
  let map f = function
    | Lam x      -> Lam (f x)
    | App (x, y) -> App(f x, f y)
    | Let (x, y) -> Let(f x, f y)
    | Annot(t, x) -> Annot(t, f x)

  let join m = function
    | Lam x -> x
    | App(x, y) -> m.join x y
    | Let(x, y) -> m.join x y
    | Annot(_, x) -> x
end

module Syntax = Abt(Lambda)

module Bidir = struct
  open Lambda
  open Syntax
  type ctx = (var * tp) list

  let is_synth = function
    | Tm (Lam _) | Tm (Let (_, _)) -> false
    | _ -> true
  let is_check e = not(is_synth e)

  let unabs e =
    match out e with
    | Abs(x, e) -> (x, e)
    | _ -> assert false

  let rec check ctx e tp =
    match out e, tp with
    | Tm (Lam t), Arrow(tp1, tp') ->
      let (x, e') = unabs t in
      check ((x, tp1) :: ctx) e' tp'
    | Tm (Lam _), _               -> failwith "expected arrow type"
    | Tm (Let(e', t)), _ ->
      let (x, e'') = unabs t in
      let tp1 = synth ctx e' in
      check ((x, tp1) :: ctx) e'' tp
    | body, _ when is_synth body ->
      if tp = synth ctx e then () else failwith "Type mismatch"
    | _ -> assert false

  and synth ctx e =
    match out e with
    | Var x -> (try List.assoc x ctx with Not_found -> failwith "unbound variable")
    | Tm(Annot(tp, e)) -> let () = check ctx e tp in tp
    | Tm(App(f, e))  ->
      (match synth ctx f with
       | Arrow(tp, tp') -> let () = check ctx e tp in tp'
       | _ -> failwith "Applying a non-function!")
    | body when is_check body -> failwith "Cannot synthesize type for checking term"
    | _ -> assert false
end

File: abt.scala
// Translate https://gist.github.com/neel-krishnaswami/834b892327271e348f79
package abt
import language.higherKinds

import scalaz.Functor

// Semi-standard typeclasses.
// TODO: Refactor to use Scalaz/Cats {{{
trait Foldable[F[T]] {
  def fold[T](v: F[T])(implicit m: Monoid[T]): T
}

trait Monoid[T] {
  def unit: T
  def join(a: T, b: T): T
}

object Monoid {
  implicit def setMonoid[T] =
    new Monoid[Set[T]] {
      def unit = Set.empty
      def join(a: Set[T], b: Set[T]): Set[T] = a ++ b
    }
}

trait ABTs {
  type Name = String
  type Vars = Set[Name]
  /* XXX: Don't use I for interfaces. What's the right convention? */
  trait IAbt[Signature[T]] {
    //Concrete type used to build terms.
    sealed trait BindingTerm[T]
    case class Var[A](n: Name) extends BindingTerm[A]
    case class Abs[T](n: Name, t: T) extends BindingTerm[T]
    case class Tm[T](t: Signature[T]) extends BindingTerm[T]
    def map[A, B](f: A => B): BindingTerm[A] => BindingTerm[B]

    //Abstract type of terms
    type Term

    //Term is isomorphic to BindingTerm[Term]
    def into(t: BindingTerm[Term]): Term
    def out(t: Term): BindingTerm[Term]

    def freeVars(t: Term): Vars

    def makeVar(n: Name): Term
    def makeAbs(n: Name, body: Term): Term
    def makeTm(t: Signature[Term]): Term
    def subst(outer: Term, v: Name, inner: Term): Term
  }

  class Abt[Signature[_]](implicit FoldableSignature: Foldable[Signature], FuncSign: Functor[Signature])
      extends IAbt[Signature] {
    def map[A, B](f: A => B): BindingTerm[A] => BindingTerm[B] = {
      case Var(n) => Var(n)
      case Abs(n, body) => Abs(n, f(body))
      case Tm(t) => Tm(FuncSign.map(t)(f))
    }

    case class Term(vars: Vars, t: BindingTerm[Term])

    def into(t: BindingTerm[Term]): Term =
      t match {
        case Var(n) => makeVar(n)
        case Abs(n, body) => makeAbs(n, body)
        case Tm(t) => makeTm(t)
      }

    def out(t: Term): BindingTerm[Term] = t.t

    def makeVar(n: Name): Term = Term(Set(n), Var(n))
    def makeAbs(n: Name, body: Term): Term =
      Term(freeVars(body) - n, Abs(n, body))

    def makeTm(t: Signature[Term]): Term =
      Term(FoldableSignature.fold(FuncSign.map(t)(freeVars)), Tm(t))

    def freeVars(t: Term): Vars = t.vars

    var index = 0
    def fresh(): Name = {
      index += 1
      "x" + index
    }
    def fresh(baseName: Name, vars: Vars): Name =
      if (vars contains baseName)
        fresh(baseName + "'", vars)
      else
        baseName

    def subst(outer: Term, v: Name, inner: Term): Term =
      subst(outer, v, inner, true)

    //This is a basic substitution with quadratic complexity.
    def subst(outer: Term, v: Name, inner: Term, preRename: Boolean): Term =
      out(outer) match {
        case Var(name) if v == name => inner
        case Tm(body) =>
          makeTm(FuncSign.map[Term, Term](body)(x => subst(x, v, inner, preRename)))
        case Abs(name, body) if v != name =>
          val (name1, body1) =
            if (preRename) {
              val newName = fresh(name, freeVars(body) ++ freeVars(outer))
              val newBody = subst(body, name, makeVar(newName), false)
              (newName, newBody)
            } else {
              //We're replacing
              (name, body)
            }
          val body2 = subst(body1, v, inner, preRename)
          makeAbs(name1, body2)
        case _ => //For when guards fail
          outer
      }
  }

  trait SimpleType
  case object Base extends SimpleType
  case class Arrow(t1: SimpleType, t2: SimpleType) extends SimpleType

  sealed trait TLambda[T]
  case class Abs[T](t: T) extends TLambda[T]
  case class App[T](t1: T, t2: T) extends TLambda[T]
  case class Let[T](t1: T, t2: T) extends TLambda[T]
  case class Annot[T](tp: SimpleType, t: T) extends TLambda[T]

  implicit val lambdaSig: Functor[TLambda] with Foldable[TLambda] =
    new Functor[TLambda] with Foldable[TLambda] {
      def map[A, B](fa: TLambda[A])(f: A => B): TLambda[B] = fa match {
        case Abs(t) => Abs(f(t))
        case Annot(tp, t) => Annot(tp, f(t))
        case App(t1, t2) => App(f(t1), f(t2))
        case Let(t1, t2) => Let(f(t1), f(t2))
      }

      def fold[T](v: TLambda[T])(implicit m: Monoid[T]): T =
        v match {
          case Abs(t) => t
          case Annot(_, t) => t
          case App(t1, t2) => m.join(t1, t2)
          case Let(t1, t2) => m.join(t1, t2)
        }
    }

  val lambdaAbt: IAbt[TLambda] = new Abt
}
